### PR TITLE
Provide option for specifying the profiler runtime

### PR DESCRIPTION
**Pull request by eggyal**
Currently, if `-Zinstrument-coverage` is enabled, the target is linked
against the `library/profiler_builtins` crate (which pulls in LLVM's
compiler-rt runtime).

This option enables backends to specify an alternative runtime crate for
handling injected instrumentation calls.

### CODE REVIEW

File: compiler/rustc_interface/src/tests.rs [hunk 1]
```diff
@@ -640,6 +640,7 @@ fn test_debugging_options_tracking_hash() {
     untracked!(no_interleave_lints, true);
     untracked!(no_leak_check, true);
     untracked!(no_parallel_llvm, true);
+    untracked!(profiler_runtime, None);
```
Comment (Aaron1011) [inline compiler/rustc_interface/src/tests.rs:643]: Why did you make this `untracked`?

Comment (eggyal) [reply compiler/rustc_interface/src/tests.rs:643]: I, perhaps wrongly, thought that if the profiler runtime is changed from one crate to another then the new dependency will necessarily trigger that crate to be built and linked in place of the old one.   But now you've raised it, I guess the new dependency won't even be found?

Comment (eggyal) [reply compiler/rustc_interface/src/tests.rs]: I reverted to tracked.

File: compiler/rustc_metadata/src/creader.rs [hunk 2]
```diff
@@ -769,27 +769,31 @@ impl<'a> CrateLoader<'a> {
...
+                info!("loading profiler");
+
+                if name == sym::profiler_builtins
+                    && self.sess.contains_name(&krate.attrs, sym::no_core)
+                {
+                    self.sess.err(
+                        "`profiler_builtins` crate (required by compiler options) \
+                                is not compatible with crate attribute `#![no_core]`",
+                    );
+                }
```
Comment (jackh726) [inline compiler/rustc_metadata/src/creader.rs:789]: The nesting here is starting to get a messy. Maybe just early return for the first if and if `profiler_runtime` is `None`?

**Review by jackh726 — APPROVED** [review]
So, is there *currently* any work being held up elsewhere by this? I.e. is there another profiler runtime ready?

I'm also not sure if this deserves an MCP. I'm not sure how used this option is?

There might be someone better to review this then me (someone more familiar in this area). That being said, the code changes here look mostly fine (modulo one nit). Are there any tests that can be added here?

**Comment by jackh726** [comment]
@eggyal do you know someone who might be a better reviewer? I'm basically okay approving this, if
1) You don't know of a better reviewer
2) `no_profiler_runtime` is in fact probably not used really anywhere (otherwise this probably needs an MCP)
3) There aren't any tests or such we can add

**Comment by eggyal** [comment]
I have an alternative profiler runtime for an under-development incremental test runner that uses cg_clif, but it's not ready no.

@bjorn3 do you have r+ privileges?  Would you be an appropriate reviewer for this?  Else perhaps @tmandry or @wesleywiser given that the profiler runtime is part of the coverage instrumentation that I think they oversaw?

**Comment by wesleywiser** [comment]
This seems fine to me but let's cc @Amanieu who added `-Zno-profiler-runtime`.

**Comment by Amanieu** [comment]
I added that flag for [minicov](https://github.com/Amanieu/minicov/) which provides a `no_std` profiling runtime (based on LLVM's).

File: compiler/rustc_metadata/src/creader.rs (hunk 2 above)
Comment (eggyal) [reply compiler/rustc_metadata/src/creader.rs]: Thanks, have done!

**Comment by jackh726** [comment]
Okay, seems fine to me. @Amanieu I suppose that the docs for minicov will have to be updated to use `-Zprofiler-runtime=None` (that's how you would specify no runtime right?) instead of `-Zno-profiler-runtime`).

@bors r+

**Pull request merged**